File: src/main/scala/com/sripe/spark/scala/WordCount.scala
package com.sripe.spark.scala

/**
  * Created by shiyu on 6/22/2016.
  */

import org.apache.spark._

object WordCount {
  def main(args: Array[String]) {
    val conf = new SparkConf().setMaster("spark://learnhadoopnode:7077").setAppName("WordCount").setJars(Seq("G:\\AndroidStudioProjects\\SparkLearningDemos\\com.sripe.spark.jar"))
    val sc = new SparkContext(conf)
    val file = sc.textFile("hdfs://learnhadoopnode:9000/user/shiyu/spark/test")
    //val wordContains = file.flatMap(f => f.split(" ")).map(f => (f, 1)).reduceByKey(_ + _).sortByKey()
    //sort the wordcounts by value
    val wordContains = file.flatMap(f => f.split(" ")).map(f => (f, 1)).reduceByKey(_ + _).sortBy(b => -b._2)
    //val wordContains = file.flatMap(f => f.split(" ")).map(f => (f, 1)).reduceByKey(_ + _).map(i => i.swap).sortByKey(false).map(i => i.swap)
    val wordCollect = wordContains.collect()
    wordCollect.foreach(println(_))
  }
}

File: src/main/scala/com/sripe/spark/java/WordCount.java
package com.sripe.spark.java;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.storage.StorageLevel;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Calendar;
import java.util.List;

/**
 * Created by shiyu on 6/24/2016.
 */
public class WordCount {
    public static void main(String[] args) {
        long startTime, endTime;
        startTime=Calendar.getInstance().getTimeInMillis();
        SparkConf sparkConf = new SparkConf().setAppName("WordCount With JAVA")
                .setMaster("spark://learnhadoopnode:7077")
                .setJars(new String[]{"G:\\AndroidStudioProjects\\SparkLearningDemos\\com.sripe.spark.jar"});
        JavaSparkContext jsc = new JavaSparkContext(sparkConf);
        JavaRDD<String> distFile = jsc.textFile("hdfs://learnhadoopnode:9000/user/shiyu/spark/test");
        JavaRDD<String> words = distFile.flatMap(new FlatMapFunction<String, String>() {
            public Iterable<String> call(String s) throws Exception {
                return Arrays.asList(s.split(" "));
            }
        }).persist(StorageLevel.MEMORY_ONLY());

        /*JavaPairRDD<String, Integer> counts = words.mapToPair(new PairFunction<String, String, Integer>() {
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s, 1);
            }
        }).reduceByKey(new Function2<Integer, Integer, Integer>() {
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        }).sortByKey();*/
        JavaPairRDD<String, Integer> counts = words.mapToPair(new PairFunction<String, String, Integer>() {
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s, 1);
            }
        }).aggregateByKey(0, new Function2<Integer, Integer, Integer>() {
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        }, new Function2<Integer, Integer, Integer>() {
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1+v2;
            }
        }).sortByKey();
        List<Tuple2<String, Integer>> sample = counts.collect();
        for (Tuple2<?, ?> item :
                sample) {
            System.out.println(item.toString());
        }
        endTime=Calendar.getInstance().getTimeInMillis();
        System.out.println("Time elapsed:"+(endTime-startTime)/1000);
    }
}
